File: python/algorithms/tree/segmenttree.py
from sys import maxsize

minsize = -99999

def maxquery(segtree, qlo, qhi, lo, hi, pos):
    if qlo <= lo and qhi >= hi:
        return segtree[pos]
    if qlo > hi or qhi < lo:
        return minsize
    mid = (lo + hi) // 2
    return max(
        maxquery(segtree, qlo, qhi, lo, mid, 2 * pos + 1),
        maxquery(segtree, qlo, qhi, mid + 1, hi, 2 * pos + 2))

def construct(arr, segtree, lo, hi, pos):
    if lo == hi:
        segtree[pos] = arr[lo]
        return
    mid = (lo + hi) // 2
    construct(arr, segtree, lo, mid, 2 * pos + 1)
    construct(arr, segtree, mid + 1, hi, 2 * pos + 2)
    segtree[pos] = max(segtree[2 * pos + 1], segtree[2 * pos + 2])

if __name__ == '__main__':
    A = [-1, 0, 3, 2, 5]
    tree = [minsize] * 2 * (len(A))
    construct(A, tree, 0, len(A) - 1, 0)
    print(maxquery(tree, 2, 4, 0, 4, 0))
    print(tree)

File: c++/datastructures/segmenttree.cpp
#include <iostream>
#include <assert.h>
#include <math.h>

#define MAX_VALUE 2100000

using namespace std;

int minquery(int segtree[], int qlo, int qhi, int lo, int hi, int pos)
{
    if (qlo <= lo && qhi >= hi) {
        return segtree[pos];
    }
    if (qlo > hi || qhi < lo) {
        return MAX_VALUE;
    }
    int mid = lo + (hi - lo) / 2;
    return min(
        minquery(segtree, qlo, qhi, lo, mid, 2 * pos + 1),
        minquery(segtree, qlo, qhi, mid + 1, hi, 2 * pos + 2));
}

void construct(int input[], int segtree[], int lo, int hi, int pos)
{
    if (lo == hi) {
        segtree[pos] = input[lo];
        return;
    }
    int mid = lo + (hi - lo) / 2;
    construct(input, segtree, lo, mid, 2 * pos + 1);
    construct(input, segtree, mid + 1, hi, 2 * pos + 2);
    segtree[pos] = min(segtree[2 * pos + 1], segtree[2 * pos + 2]);
}

int main()
{
    int freq[] {2, 4, 5, 2, 1, 9, 10, 7, 6, 2};
    int n = sizeof(freq) / sizeof(freq[0]);
    int size = 2 * (int)pow(2, ceil(log2(n))) - 1;
    int segtree[size];
    construct(freq, segtree, 0, n - 1, 0);    
    cout << minquery(segtree, 0, 5, 0, n - 1, 0) << endl;
    return 0;
}

